### PR TITLE
Fix L3 timestamps

### LINKED ISSUE #2277: Higher or negative latency using 2.4.3 (time - sent_time)
#### Brief description

Latency calculated is 10 times higher than 2.4.0, and sometimes, latency is negative.

Looks like https://github.com/secdev/scapy/issues/1791 but with negative values now.

The test is done in TCP.

It does not seem there is a mismatch between receive and sent packets as SEQ/ACK and SPORT/DPORT matches.

#### Environment

Scapy version: Scapy 2.4.3
Python version: 3.6 & 3.7
Operating System: centos 7.4 & arch linux

Tests are done in a LAN with stable machine, and also in production in datacenters.

#### How to reproduce

```from scapy.all import *

dest = "192.168.0.1"

pkt = []
pkt_rst = []
for i in range(1, 100):
    a = IP(dst=dest) / TCP(flags="S", seq=i, sport=65000+i, dport=55556)
    b = IP(dst=dest) / TCP(flags="R", seq=i, sport=65000+i, dport=55556)
    pkt.append(a)
    pkt_rst.append(b)

ans, unans = sr(pkt, filter="host {0}".format(dest), inter=0, timeout=1)
send(pkt_rst, inter=0)

print("scapy version: {}".format(conf.version))

for pkt in ans:
    sent = pkt[0]
    received = pkt[1]
    res = (received.time - sent.sent_time) * 1000
    print(res)
    if res < 0:
        breakpoint()
```

#### Actual result

Using 2.4.0:
```sudo python test.py
Begin emission:
.****************************************************Finished sending 99 packets.
***********************************************
Received 100 packets, got 99 answers, remaining 0 packets
...................................................................................................
Sent 99 packets.
scapy version: 2.4.0
0.48351287841796875
0.31113624572753906
0.46896934509277344
....
0.27561187744140625
0.3025531768798828
0.38909912109375
0.3840923309326172
0.3809928894042969
0.2892017364501953
```

Using 2.4.3:
```
sudo python ./test.py
Begin emission:
.**************************************************************************************************Finished sending 99 packets.
*
Received 100 packets, got 99 answers, remaining 0 packets
...................................................................................................
Sent 99 packets.
scapy version: 2.4.3
3.6242008209228516
2.803325653076172
3.2286643981933594
7.400989532470703
12.633323669433594
13.63062858581543
15.126228332519531
11.01994514465332
24.735450744628906
24.316787719726562
23.879051208496094
23.487329483032227
22.953510284423828
22.533655166625977
22.108078002929688
40.10128974914551
39.75367546081543
39.3528938293457
38.92111778259277
38.49363327026367
38.06662559509277
37.683963775634766
36.977529525756836
36.301374435424805
34.806013107299805
34.380197525024414
34.37399864196777
34.02996063232422
33.62083435058594
33.17570686340332
32.75179862976074
32.33003616333008
24.41573143005371
23.94843101501465
23.525714874267578
23.09560775756836
22.677898406982422
22.095203399658203
21.67963981628418
21.364927291870117
20.95651626586914
20.496606826782227
20.053863525390625
19.638776779174805
19.222021102905273
18.764257431030273
18.578290939331055
18.302440643310547
19.684314727783203
...
-0.3478527069091797
[0] > /home/kevin/tmp/scapy-test/test.py(18)<module>()
-> for pkt in ans:
(Pdb++) sent
<IP  frag=0 proto=tcp dst=192.168.18.71 |<TCP  sport=65067 dport=55556 seq=67 flags=S |>>
(Pdb++) received
<IP  version=4 ihl=5 tos=0x0 len=40 id=0 flags=DF frag=0 ttl=64 proto=tcp chksum=0x94e4 src=192.168.18.71 dst=192.168.18.84 |<TCP  sport=55556 dport=65067 seq=0 ack=68 dataofs=5 reserved=0 flags=RA window=0 chksum=0x3270 urgptr=0 |<Padding  load='\x00\x00\x00\x00\x00\x00' |>>>

```

### LINKED ISSUE #2423: Packet Timestamps being overwritten - Identified a Solution
#### Brief description
I have installed the latest commit of scapy with the "Rework Timestamp Detection" commit: 
[#2401](https://github.com/secdev/scapy/commit/90c8d82070a9edaf0bbfe53b00db8dcf13154eed)

I have determined that the packet received timestamp is being overwritten.
#### Environment

-   Scapy version: `v2.4.3-237.g90c8d820`
-   Python version: `e.g. 3.7`
-   Operating System: `e.g. Fedora 30 (NST 30)`

#### How to reproduce

Use the following script to send four (4) TCP sync packets to your default gateway. On a typical LAN you should get sub-milliseconds round-trip-time results. Substitute your default gateway for the 'dest' variable.

```
#
# Scapy script to compute Round Trip Times (rrt)...
from scapy.all import *

dest = "Your Gateway IPv4 Address"

pkt = []
pkt_rst = []
for i in range(1, 5):
    a = IP(dst=dest) / TCP(flags="S", seq=i, sport=65000+i, dport=55556)
    b = IP(dst=dest) / TCP(flags="R", seq=i, sport=65000+i, dport=55556)
    pkt.append(a)
    pkt_rst.append(b)

ans, unans = sr(pkt, filter="host {0}".format(dest), inter=0, timeout=1)
send(pkt_rst, inter=0)

print("scapy version: {}".format(conf.version))

for pkt in ans:
    sent = pkt[0]
    received = pkt[1]
    res = (received.time - sent.sent_time) * 1000
    print(res)
    if res < 0:
        breakpoint()

```
#### Actual result

My source address: 10.222.222.10 (shopper2) and my gateway: 10.222.222.1:

```
[root@shopper2 opt]# python3 ./rrt-scapy.py 
Begin emission:
Finished sending 4 packets.

Received 5 packets, got 4 answers, remaining 0 packets
....
Sent 4 packets.
scapy version: v2.4.3-237.g90c8d820
9.302377700805664
7.546901702880859
5.416631698608398
3.2846927642822266
[root@shopper2 opt]#
```
As you can see the round-trip-times are way to big.

#### The Problem Area and Solution

After much debugging I found that the received packet was being overwritten in the "L3PacketSocket" class. Normally I would submit a Pull Request. But in this case I am only solving for the Linux platform and I am uncertain if the fix will be the permanent solution.

In file "arch/linux.py"  I have preserved the received timestamp before assigning the "pkt" with the "pkt.payload" value:

Before:
```
class L3PacketSocket(L2Socket):
    desc = "read/write packets at layer 3 using Linux PF_PACKET sockets"

    def recv(self, x=MTU):
        pkt = SuperSocket.recv(self, x)
        if pkt and self.lvl == 2:
            pkt = pkt.payload
        return pkt
```
After:
```
class L3PacketSocket(L2Socket):
    desc = "read/write packets at layer 3 using Linux PF_PACKET sockets"

    def recv(self, x=MTU):
        pkt = SuperSocket.recv(self, x)
        if pkt and self.lvl == 2:
            ts = pkt.time
            pkt = pkt.payload
            pkt.time = ts
        return pkt
```

Suggestion: The "pkt.time" may could be renamed to "pkt.recv_time"

#### Expected result

After the modification is in place the following are the results:

```
[root@shopper2 opt]# python3 ./rrt-scapy.py 
Begin emission:
Finished sending 4 packets.

Received 5 packets, got 4 answers, remaining 0 packets
....
Sent 4 packets.
scapy version: v2.4.3-237.g90c8d820
0.43010711669921875
0.34427642822265625
0.28777122497558594
0.2853870391845703
[root@shopper2 opt]#
```
Now sub-millisecond round-trip-times are measured using TCP Sync packets. To verify lets do a ping to the gateway (10.222.222.1):

```
[root@shopper2 opt]# ping -c 4 -nv 10.222.222.1
PING 10.222.222.1 (10.222.222.1) 56(84) bytes of data.
64 bytes from 10.222.222.1: icmp_seq=1 ttl=64 time=0.413 ms
64 bytes from 10.222.222.1: icmp_seq=2 ttl=64 time=0.327 ms
64 bytes from 10.222.222.1: icmp_seq=3 ttl=64 time=0.339 ms
64 bytes from 10.222.222.1: icmp_seq=4 ttl=64 time=0.331 ms

--- 10.222.222.1 ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 100ms
rtt min/avg/max/mdev = 0.327/0.352/0.413/0.039 ms
[root@shopper2 opt]# 
```
As you can see the ICMP round-trip-time results agree nicely with the TCP Sync packets.

#### Scapy Multi-Traceroute Add On: mtraceroute
I will be announcing this scapy Multi-Traceroute add on feature: [mtraceroute](https://github.com/rwhalb/mtraceroute) once we get the packet timestamps corrected.

---Ron Henderson
CoAuthor of NST (Network Security Toolkit)

### LINKED ISSUE #2425: _recv_raw is breaking ISOTPNativeSockets
@gpotter2 
Hi, on some hardware in the loop tests, we discovered the following error:
```
WARNING: Socket <scapy.contrib.isotp.ISOTPNativeSocket object at 0x7319dc70> failed with 'unpack requires a buffer of 16 bytes'. It was closed.
```
We assume this error is fired from this line:
https://github.com/secdev/scapy/blob/ae6b93b9263cab4319be463527b92d1edf13ce43/scapy/supersocket.py#L109

Unfortunately, I don't know exactly whats going on here, but my ISOTPNativeSockets get closed on a receive.

### LINKED ISSUE #2429: when process has required capabilities use them
#### Things to consider

1.  Please check that you are using the latest scapy version:
Yes, I am. the latest version is[ 2.4.3](https://github.com/secdev/scapy/releases/tag/v2.4.3), here's output of $ `pip show scapy`:
```
Name: scapy
Version: 2.4.3
Summary: Scapy: interactive packet manipulation tool
Home-page: https://scapy.net
Author: Philippe BIONDI
Author-email: guillaume@valadon.net
License: GPLv2
Location: /home/x/.local/lib/python3.8/site-packages
Requires: 
Required-by: 
```

#### Brief description
tcpdump not passed ambient CAP_NET_RAW capability given to python standalone exe aot compiled with nuitka

#### Environment

-   Scapy version: 2.4.3
-   Python version: 3.8
-   Operating System: latest rolling arch linux release and updated packages as of today, output of $ `uname -a`: `Linux archiver 5.4.12-arch1-1 #1 SMP PREEMPT Tue, 14 Jan 2020 21:44:31 +0000 x86_64 GNU/Linux`

#### How to reproduce

use arping that uses tcpdump, add CAP_NET_RAW capability to python interpreter temporarly for testing or compile the script with nuitka and add the capability with setcomp to the standalone executable

#### Actual result
```
Traceback (most recent call last):
  File "dist/watch_LAN.py", line 127, in <module>
  File "dist/watch_LAN.py", line 87, in main
  File "dist/watch_LAN.py", line 68, in mac_discovery
  File "dist/scapy/layers/l2.py", line 628, in arping
  File "dist/scapy/sendrecv.py", line 503, in srp
  File "dist/scapy/arch/linux.py", line 475, in __init__
  File "dist/scapy/arch/linux.py", line 160, in attach_filter
  File "dist/scapy/arch/common.py", line 182, in compile_filter
scapy.error.Scapy_Exception: Failed to attach filter: tcpdump returned: b"tcpdump: lo: You don't have permission to capture on that device\n(socket: Operation not permitted)\n"
```

#### Expected result
tcpdump should be passed the capabilities and be able to capture raw packets

you can't set capabilities on a python script, or any script, with a shebang or not e.g bash script, however, python has compilers that can make a standalone executable that you can set capabilities on.

and you shouldn't set capabilities on interpreters:
the accepted answer on the question: [Python Scapy sniff without root](https://stackoverflow.com/questions/36215201/python-scapy-sniff-without-root)
proposes setting capabilities on the python interpreter and tcpdump, but this is not secure at all, this enables any python script to have those capabilities and anything that uses tcpdump too. the [second answer](https://stackoverflow.com/a/47982075/4178053) proposes using a helper binary that you set the capabilities on and it passes them to the interpreter it starts, but this also has the first drawback of the first answer (allowing anything ran using that helper binary to have those capabilities)

I need the security of not using root and using linux capabilities instead.

### PR DESCRIPTION
**Pull request by gpotter2**
Credits for the timestamp bug to Ron Henderson

fix #2423
fix #2425 
fix #2277
fix #2429

### CODE REVIEW

File: scapy/supersocket.py [hunk 1]
```diff
@@ -106,8 +106,12 @@ def _recv_raw(self, sock, x):
                         pkt = pkt[:12] + tag + pkt[12:]
                 elif cmsg_lvl == socket.SOL_SOCKET and \
                         cmsg_type == SO_TIMESTAMPNS:
-                    tmp = struct.unpack("iiii", cmsg_data)
-                    timestamp = tmp[0] + tmp[2] * 1e-9
+                    length = len(cmsg_data)
+                    if length == 16:  # __kernel_timespec
+                        tmp = struct.unpack("ll", cmsg_data)
+                    elif length == 8:  # timespec
+                        tmp = struct.unpack("ii", cmsg_data)
```
Comment (guedou) [inline scapy/supersocket.py]: Can you add an `else:` case to handle an invalid length? That will prevent weird errors.

File: scapy/arch/bpf/supersocket.py [hunk 2]
```diff
@@ -354,7 +354,10 @@ def recv(self, x=BPF_BUFFER_LENGTH):
         """Receive on layer 3"""
         r = SuperSocket.recv(self, x)
         if r:
-            return r.payload
+            ts = r.time
+            r = r.payload
+            r.time = ts
```
Comment (p-l-) [inline scapy/arch/bpf/supersocket.py]: No need for a temporary variable here or another variable assignment:
```suggestion
            r.payload.time = r.time
            return r.payload
```

File: scapy/arch/linux.py [hunk 3]
```diff
@@ -533,7 +533,9 @@ class L3PacketSocket(L2Socket):
     def recv(self, x=MTU):
         pkt = SuperSocket.recv(self, x)
         if pkt and self.lvl == 2:
+            ts = pkt.time
             pkt = pkt.payload
+            pkt.time = ts
```
Comment (p-l-) [inline scapy/arch/linux.py]: Same here: 
```suggestion
            pkt.payload.time = pkt.time
            return pkt.payload 
```

File: scapy/data.py [hunk 4]
```diff
@@ -30,7 +30,7 @@
 SOL_PACKET = 263
 # From asm/socket.h
 SO_ATTACH_FILTER = 26
-SO_TIMESTAMPNS = 35
+SO_TIMESTAMPNS = 35  # SO_TIMESTAMPNS_OLD: not 2038 safe
```
Comment (p-l-) [inline scapy/data.py]: What do you mean?

Comment (gpotter2) [reply scapy/data.py]: See https://github.com/torvalds/linux/blob/a2d79c7174aeb43b13020dd53d85a7aefdd9f3e5/include/uapi/asm-generic/socket.h#L109-L115

- `SO_TIMESTAMPNS_OLD = 35` returns timespecs that use long to store the date = not 2038 safe
- `SO_TIMESTAMPNS_NEW` returns timespecs that are long long
- `SO_TIMESTAMPNS` defaults for the old one

We would use the "NEW" one but it's very recent and isn't supported well yet.
https://lwn.net/Articles/776219/

**Comment by gpotter2** [comment]
Should be ready to merge

**Pull request merged**